oap-server/server-alarm-plugin/src/main/java/org/apache/skywalking/oap/server/core/alarm/provider/CompositeRuleEvaluator.java

```diff
@@ -0,0 +1,127 @@
+     * @param alarmMessages       triggered alarm messages
+     * @return
+     */
+    public List<AlarmMessage> evaluate(List<CompositeAlarmRule> compositeAlarmRules, List<AlarmMessage> alarmMessages) {
+        final List<AlarmMessage> compositeRuleMessages = new ArrayList<>();
+        ImmutableListMultimap<String, AlarmMessage> messageMap = Multimaps.index(alarmMessages, alarmMessage ->
+                Joiner.on(Const.ID_CONNECTOR).useForNull(Const.EMPTY_STRING).join(alarmMessage.getId0(), alarmMessage.getId1()));
+        for (CompositeAlarmRule compositeAlarmRule : compositeAlarmRules) {
+            String expr = compositeAlarmRule.getExpression();
+            Set<String> dependencyRules = expression.analysisInputs(expr);
+            Map<String, Object> dataContext = new HashMap<>();
+            messageMap.asMap().forEach((key, alarmMessageList) -> {
+                dependencyRules.forEach(ruleName -> dataContext.put(ruleName, false));
+                alarmMessageList.forEach(alarmMessage -> {
+                    if (dependencyRules.contains(alarmMessage.getRuleName())) {
+                        dataContext.put(alarmMessage.getRuleName(), true);
+                    }
+                });
+                Object matched = expression.eval(expr, dataContext);
+                if (matched instanceof Boolean && (Boolean) matched) {
+                    AlarmMessage headMsg = alarmMessageList.iterator().next();
+                    AlarmMessage message = new AlarmMessage();
+                    message.setOnlyAsCondition(false);
+                    message.setScopeId(headMsg.getScopeId());
+                    message.setScope(headMsg.getScope());
+                    message.setName(headMsg.getName());
+                    message.setId0(headMsg.getId0());
+                    message.setId1(headMsg.getId1());
+                    message.setStartTime(System.currentTimeMillis());
+                    message.setRuleName(compositeAlarmRule.getAlarmRuleName());
+                    String alarmMessage = formatMessage(message, compositeAlarmRule.getMessage());
+                    message.setAlarmMessage(alarmMessage);
+                    compositeRuleMessages.add(message);
+                }
+            });
+        }
+        return compositeRuleMessages;
+    }
+
+    private String formatMessage(AlarmMessage alarmMessage, String message) {
+        return messageFormatterCache.computeIfAbsent(message, AlarmMessageFormatter::new).format(new MetaInAlarm() {
+            @Override
+            public String getScope() {
+                return null;
+            }
+
+            @Override
+            public int getScopeId() {
+                return alarmMessage.getScopeId();
+            }
+
+            @Override
+            public String getName() {
+                return alarmMessage.getName();
+            }
+
+            @Override
+            public String getMetricsName() {
+                return null;
```

Comment:
The metrics name should be filled as the expression literal text.

Reply:
done